proposals/inline-classes.md

```diff
@@ -0,0 +1,163 @@
+- Inline class must be a toplevel class 
+
+### Other restrictions
+
+The following restrictions are related to the usages of inline classes:
+
+- Referential equality (`===`) is prohibited for inline classes
+- vararg of inline class type is prohibited
+```
+inline class Foo(val s: String)
+
+fun test(vararg foos: Foo) { ... } // should be an error  
+```
+
+## Java interoperability  
+
+Each inline class has its own wrapper, which is boxed/unboxed by the same rules as for primitive types.
+Basically, rule for boxing can be formulated as follows: inline class is boxed when it is used as another type.
+Unboxed inline class is used when value is statically known to be inline class.
+
+Examples:
+```
+interface I
+
+inline class Foo(val i: Int) : I
+
+fun asInline(f: Foo) {}
+fun <T> asGeneric(x: T) {}
+fun asInterface(i: I) {}
+fun asNullable(i: Foo?) {}
+
+fun <T> id(x: T): T = x
+
+fun test(f: Foo) {
+    asInline(f)
+    asGeneric(f) // boxing
+    asInterface(f) // boxing
+    asNullable(f) // boxing
+    
+    val c = id(f) // boxing/unboxing, c is unboxed
+}
+```
+
+Since boxing doesn't have side effects as is, it's possible to reuse various optimizations that are done for primitive types.
+
+### Type mapping on JVM
+
+Depending on the underlying type of inline class and its declaration site, inline class type can be mapped to the underlying type or to the 
+wrapper type.  
+
+
+| Underlying Type \ Declaration Site | Not-null | Nullable | Generic |
+| --------------------------------------------- | --------------- | ---------------- | ---------------- |
+| **Reference type** | Underlying type | Underlying type | Wrapper type |
+| **Nullable reference type** | Underlying type  | Wrapper type | Wrapper type |
+| **Primitive type** | Underlying type  | Wrapper type | Wrapper type |
+| **Nullable primitive type** | Underlying type  | Wrapper type | Wrapper type |
+
+For example, from this table it follows that nullable inline class which is based on reference type is mapped to the underlying type:
+```
+inline class Name(val s: String)
+
+fun foo(n: Name?) { ... } // `Name?` here is mapped to String
+```
+
+## Methods from `kotlin.Any`
+
+Inline classes are indirectly inherited from `Any`, i.e. they can be assigned to a value of type `Any`, but only through boxing.
+
+Methods from `Any` (`toString`, `hashCode`, `equals`) can be useful for a user-defined inline classes and therefore should be customizable. 
+Methods `toString` and `hashCode` can be overridden as usual methods from `Any`. For method `equals` we're going to introduce new operator 
+that represents "typed" `equals` to avoid boxing for inline classes:
+```
+inline class Foo(val s: String) {
+    operator fun equals(other: Foo): Boolean { ... }
+}
+```
+
+Compiler will generate original `equals` method that is delegated to the typed version.  
+
+By default, compiler will automatically generate `equals`, `hashCode` and `toString` same as for data classes.
```

Comment:
BTW, are data inline classes allowed?

Reply:
Well, they can be allowed... I'm not sure it makes sense, I'll add a note

Reply:
Related - is there a reason sealed classes can't support inlining as long as every subclass is also inline?

Reply:
@hzsweers Since there's no instance, how would you do `is` checks at runtime for use in a `when` or boolean expression?

Reply:
I would assume that they would work the same way `is` checks would work in normal use (I didn't see that listed as a listed caveat and assumed that meant it was supported, though admittedly I don't see how)

Reply:
I expect `is` check to not be supported on inline classes since they are identity-less, as said [here](https://github.com/Kotlin/KEEP/pull/103#discussion_r179085704)

Reply:
@LouisCAD Could you please elaborate? Type check should work the same as it works for primitives.
`<expr> is InlineClassType`
There are two options: 
- type of `expr` is statically known to be an inline class type, then type check is trivial
- type of `expr` is statically unknown, so if it's actually of an inline class type then it should be boxed, therefore we do type check for boxed type

Reply:
@LouisCAD Yes, same as for primitives

Reply:
We can this problem in the [issue](https://github.com/Kotlin/KEEP/issues/104). We had some design of what you thinking :) 
But it has too much restrictions, you can check [opaque](https://docs.scala-lang.org/sips/opaque-types.html) types in Scala, for example